arroyo-types/src/api.rs

```diff
@@ -0,0 +1,74 @@
+use serde::{Deserialize, Serialize};
+use utoipa::ToSchema;
+
+#[derive(Serialize, Deserialize, Clone, ToSchema, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct PipelinePost {
+    pub name: String,
+    pub definition: Option<String>,
+    pub udfs: Option<Vec<Udf>>,
+}
+
+#[derive(Serialize, Deserialize, Clone, ToSchema, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct Pipeline {
+    pub id: String,
+    pub name: String,
+    pub definition: Option<String>,
+    pub udfs: Vec<Udf>,
+}
+
+#[derive(Serialize, Deserialize, Clone, ToSchema, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct JobPost {
+    pub pipeline_id: String,
+    pub running_desired: Option<bool>,
+    pub tasks: Option<u64>,
+}
+
+#[derive(Serialize, Deserialize, Clone, ToSchema, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct Job {
+    pub id: String,
+    pub pipeline_id: String,
+    pub running_desired: bool,
+    pub state: String,
+    pub run_id: u64,
+    pub start_time: Option<u64>,
+    pub finish_time: Option<u64>,
+    pub tasks: Option<u64>,
+    pub failure_message: Option<String>,
+}
+
+#[derive(Serialize, Deserialize, Clone, ToSchema, Debug, FromPrimitive)]
+#[serde(rename_all = "camelCase")]
+pub enum UdfLanguage {
+    Rust = 0,
+}
+
+#[derive(Serialize, Deserialize, Clone, ToSchema, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct Udf {
+    pub language: UdfLanguage,
+    pub definition: String,
+}
+
+#[derive(Serialize, Deserialize, Clone, ToSchema, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct PipelineCollection {
+    pub total: i32,
+    pub page: i32,
```

Comment:
What do you think about using cursors instead of pages (e.g., in https://stripe.com/docs/api/pagination)?

Page-based pagination has the issue that the data may change in between requests. So I request `/v1/jobs` and get back:


```
{
  "page": 0,
  "pages": 2,
  "items": [
    1, 2, 3, 4, 5
  ]
}
```

then I want the next page, so I request `/v1/jobs?page=1`. But in the meantime, record 2 has been deleted so I end up with 

```
{
  "page": 0,
  "pages": 2,
  "items": [
    7, 8, 9, 10, 11
  ]
}
```

(never seeing 6).

Cursor-based APIs address this by giving you a consistent position the database to query from so that you can miss records due to additions/deletions. I think it's also a bit more user friendly, since it means the user doesn't have to calculate the next page, they can just request the cursor.

Reply:
Since we're using these endpoints for a UI we probably want to have a pagination component that lets users jump to a specific page, and unfortunately that's not possible with cursor-based pagination. Practically I don't think that that race condition is a big issue because the UI will know to refresh the data when certain api calls are made (e.g. if a deletion occurs then SWR will automatically and immediately refresh the collection).

Reply:
That's true, though I'm not sure how useful jumping to a particular page is in this context. The APIs will also be used by developers directly to build their own integrations and issues like this can make it hard to build those things reliably.

Reply:
Ok I've changed it to cursor based.